src/compiler/parser.ts

```diff
@@ -8144,26 +8140,42 @@ namespace ts {
                                 }
                                 break;
                             case SyntaxKind.BacktickToken:
-                                if (state === JSDocState.SavingBackticks) {
-                                    state = JSDocState.SavingComments;
-                                }
-                                else {
-                                    state = JSDocState.SavingBackticks;
-                                }
+                                pushState(state === JSDocState.SavingBackticks ? JSDocState.SavingComments : JSDocState.SavingBackticks);
                                 pushComment(scanner.getTokenText());
                                 break;
                             case SyntaxKind.AsteriskToken:
-                                if (state === JSDocState.BeginningOfLine) {
+                                // if there are two ** after a newline, considering it as a pure comment
+                                if (state === JSDocState.BeginningOfLine && lookAhead(() => nextTokenJSDoc() !== SyntaxKind.AsteriskToken)) {
+                                    /*
+                                    * @param x
+                                    * in the first line
+                                    */
+                                    // to check tag starting newline without comments like above
+                                    // it needs to actually encounter a newline
+                                    if (states.length === 1 && lookAhead(() => nextTokenJSDoc() !== SyntaxKind.NewLineTrivia)) {
+                                        removeTrailingWhitespace(comments);
+                                    }
                                     // leading asterisks start recording on the *next* (non-whitespace) token
-                                    state = JSDocState.SawAsterisk;
+                                    pushState(JSDocState.SawAsterisk);
                                     indent += 1;
                                     break;
                                 }
                                 // record the * as a comment
                                 // falls through
                             default:
+                                /*

```

Comment:
Can you explain what the parser changes do and why they're needed?

From reading the code, it looks like `states` is added to know when to call `removeTrailingWhitespace`, and the change in `tryParseTypeExpression` now doesn't skip whitespace when a parse fails, but I don't understand the intent of shuffling whitespace parsing around.

Reply:
like you mentioned
- `tryParseTypeExpression` doesn't skip whitespace when a parse fails
```js
/*
 * @param b  // <- there is no type expression, it rollback to this position
 * 
 * something here
 */
```
- `parseTagComments` only remove whitespace for these two cases
/*
  * @param x
  * in the first line
  */
/*
 * @param x
still in the first line
 */
we can not do it at the end like we are currently doing since it is too late because there is a case with `{@link}` will reset comments 

I don't really get what you mean by `I don't understand the intent of shuffling whitespace parsing around`, can you explain about it? 😄